### DISCUSSION TITLE
[Support] 升级midway后，组件中涉及"node:url"，无法支持node14环境

### DISCUSSION THREAD

**Discussion by minefujiko9293** [midwayjs/midway#4001] [Q&A]
### Describe the problem（描述问题）

node版本：14.16.0
```
"dependencies": {
    "@midwayjs/bootstrap": "^3.16.6",
    "@midwayjs/core": "^3.16.2",
    "@midwayjs/cron": "^3.16.5",
    "@midwayjs/cross-domain": "^3.16.8",
    "@midwayjs/decorator": "^3.16.2",
    "@midwayjs/info": "^3.16.8",
    "@midwayjs/jwt": "^3.16.8",
    "@midwayjs/koa": "^3.16.8",
    "@midwayjs/logger": "^3",
    "@midwayjs/redis": "^3.16.5",
    "@midwayjs/typeorm": "3.16.5",
    "@midwayjs/upload": "^3.16.8",
    "@midwayjs/validate": "^3.16.8",
    "crypto-js": "^4.2.0",
    "dayjs": "^1.11.10",
    "lodash": "^4.17.21",
    "mime-types": "^2.1.35",
    "sqlite3": "^5.1.7",
    "typeorm": "^0.3.20",
    "ulid": "^2.3.0"
  },
  "devDependencies": {
    "@midwayjs/mock": "^3.16.5",
    "@types/crypto-js": "^4.2.2",
    "@types/jest": "^29.2.0",
    "@types/lodash": "^4.14.202",
    "@types/mime-types": "^2.1.4",
    "@types/node": "14",
    "cross-env": "^6.0.0",
    "jest": "^29.2.2",
    "mwts": "^1.3.0",
    "mwtsc": "^1",
    "pm2": "^5.3.1",
    "ts-jest": "^29.0.3",
    "typescript": "~4.8.0"
  }
```

[下午4:11:14] Found 0 errors. Watching for file changes.

(node:24652) UnhandledPromiseRejectionWarning: Error: Cannot find module 'node:url'
Require stack:
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\glob@10.4.5\node_modules\glob\dist\commonjs\glob.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\glob@10.4.5\node_modules\glob\dist\commonjs\index.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\util\DirectoryExportedClassesLoader.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\connection\ConnectionMetadataBuilder.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\data-source\DataSource.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\connection\ConnectionManager.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\globals.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\typeorm@0.3.20_sqlite3@5.1.7\node_modules\typeorm\index.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+typeorm@3.16.5\node_modules\@midwayjs\typeorm\dist\decorator.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+typeorm@3.16.5\node_modules\@midwayjs\typeorm\dist\index.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\dist\configuration.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+core@3.16.2\node_modules\@midwayjs\core\dist\common\fileDetector.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+core@3.16.2\node_modules\@midwayjs\core\dist\service\slsFunctionService.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+core@3.16.2\node_modules\@midwayjs\core\dist\index.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+mock@3.16.5\node_modules\@midwayjs\mock\dist\creator.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+mock@3.16.5\node_modules\@midwayjs\mock\dist\index.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\@midwayjs+mock@3.16.5\node_modules\@midwayjs\mock\app.js
- D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\mwtsc@1.11.2\node_modules\mwtsc\lib\wrap.js      
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (D:\workspace_ssd\mw__zsjj_cgs_qjtx\node_modules\.pnpm\glob@10.4.5\node_modules\glob\src\glob.ts:3:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:24652) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either 
by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:24652) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.      

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)     
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.16.6)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)       
✓ @midwayjs/casbin-typeorm-adapter(not installed)     
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.16.2)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(3.16.5)
✓ @midwayjs/cross-domain(3.16.8)
✓ @midwayjs/decorator(3.16.2)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.16.8)
✓ @midwayjs/jwt(3.16.8)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.16.5)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)       
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(3.16.5)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)      
✓ @midwayjs/tablestore(not installed)   
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)       
✓ @midwayjs/typegoose(not installed)    
✓ @midwayjs/typeorm(3.16.5)
✓ @midwayjs/upload(3.16.8)
✓ @midwayjs/validate(3.16.8)
✓ @midwayjs/version(not installed)      
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)     
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)      
✓ @midwayjs/koa(3.16.8)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by minefujiko9293** [comment]
经过lock文件对比，是typeorm版本造成的，写死typeorm: "0.3.17"解决

**Reply by czy88840616** [reply]
：）

**Comment by waitingsong** [comment]
node v18 都已经过了维护期，还是尽量升级到新版吧

**Reply by czy88840616** [reply]
我在考虑新版本从 18 开始了。。。

**Reply by minefujiko9293** [reply]
嘻嘻，主要文档里面写了支持node>14.
我又是做政府项目的。
资源操作系统也不支持高版本node

**Reply by czy88840616** [reply]
我说的是下一个大版本，当前版本运行时支持从 12 开始不会变的。

**Reply by waitingsong** [reply]
> 我在考虑新版本从 18 开始了。。。

midway v4 就直接从 nodejs20 开始吧。距离 nodejs 22 发布也就三个月，mw4 升级迁移再有一段时间，nodejs18 直接 EOL 了～

**Reply by waitingsong** [reply]
> 嘻嘻，主要文档里面写了支持node>14. 我又是做政府项目的。 资源操作系统也不支持高版本node

win7 支持 nodejs 最高好像是 v16.

**Reply by minefujiko9293** [reply]
>  我说的是下一个大版本，当前版本运行时支持从 12 开始不会变的。

V4有什么大动作吗 有没有剧透

**Reply by czy88840616** [reply]
https://github.com/midwayjs/midway/discussions/3641 这里有一些
